examples/embeddable_examples/public/book/book_embeddable.tsx

```diff
@@ -0,0 +1,123 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import React from 'react';
+import ReactDOM from 'react-dom';
+import { Subscription } from 'rxjs';
+import {
+  Embeddable,
+  EmbeddableInput,
+  IContainer,
+  EmbeddableOutput,
+  SavedObjectEmbeddableInput,
+  AttributeService,
+} from '../../../../src/plugins/embeddable/public';
+import { BookSavedObjectAttributes } from '../../common';
+import { BookEmbeddableComponent } from './book_component';
+
+export const BOOK_EMBEDDABLE = 'book';
+export type BookEmbeddableInput = BookByValueInput | BookByReferenceInput;
+export interface BookEmbeddableOutput extends EmbeddableOutput {
+  hasMatch: boolean;
+  attributes: BookSavedObjectAttributes;
+}
+
+interface BookInheritedInput extends EmbeddableInput {
+  search?: string;
+}
+
+export type BookByValueInput = { attributes: BookSavedObjectAttributes } & BookInheritedInput;
+export type BookByReferenceInput = SavedObjectEmbeddableInput & BookInheritedInput;
+
+/**
+ * Returns whether any attributes contain the search string.  If search is empty, true is returned. If
+ * there are no savedAttributes, false is returned.
+ * @param search - the search string
+ * @param savedAttributes - the saved object attributes for the saved object with id `input.savedObjectId`
+ */
+function getHasMatch(search?: string, savedAttributes?: BookSavedObjectAttributes): boolean {
+  if (!search) return true;
+  if (!savedAttributes) return false;
+  return Boolean(
+    (savedAttributes.author && savedAttributes.author.match(search)) ||
+      (savedAttributes.title && savedAttributes.title.match(search))
+  );
+}
+
+export class BookEmbeddable extends Embeddable<BookEmbeddableInput, BookEmbeddableOutput> {
+  public readonly type = BOOK_EMBEDDABLE;
+  private subscription: Subscription;
+  private node?: HTMLElement;
+  private savedObjectId?: string;
+  private attributes?: BookSavedObjectAttributes;
+
+  constructor(
+    initialInput: BookEmbeddableInput,
+    private attributeService: AttributeService<
+      BookSavedObjectAttributes,
+      BookByValueInput,
+      BookByReferenceInput
+    >,
+    {
+      parent,
+    }: {
+      parent?: IContainer;
+    }
+  ) {
+    super(initialInput, {} as BookEmbeddableOutput, parent);
+
+    this.subscription = this.getInput$().subscribe(async () => {
+      const savedObjectId = (this.getInput() as BookByReferenceInput).savedObjectId;
+      const attributes = (this.getInput() as BookByValueInput).attributes;
+      if (this.attributes !== attributes || this.savedObjectId !== savedObjectId) {
+        this.savedObjectId = savedObjectId;
```

Comment:
I don't understand this. If there was a change in attributes, why change a savedObjectId? Also, should there be a deep equal for attributes here?

Reply:
This line is also responsible for setting the savedObjectId to undefined when the embeddable is by value i.e. when attributes is not undefined. So it made sense to me to just update it either way!